### DISCUSSION TITLE
Ability to add description to custom editor settings

### DISCUSSION THREAD

**Discussion by starcin** [godotengine/godot-proposals#8224] [Editor]
Currently, it is possible to add some information to a custom setting to the editor settings panel is possible by using `add_property_info` with a `Dictionary` like this:
![image](https://github.com/godotengine/godot-proposals/assets/62737425/b71eab32-9779-4adb-9a36-9e3bc94ec24c)

But it results with "No description" when you hover the setting:
![image](https://github.com/godotengine/godot-proposals/assets/62737425/d56f6855-33ef-421f-a08e-999490558c02)

What I am proposing is adding a `description` field to the information dictionary:
![image](https://github.com/godotengine/godot-proposals/assets/62737425/cd70c90d-6ab9-492f-9b02-316786f05b50)

**Comment by asciifaceman** [comment]
I'm running into this now developing an addon and I was kind of shocked that this wasn't already a thing. There doesn't seem to be an API for managing descriptions in the actual project settings code - my guess is the docs system is used to lift those out of the documentation based on config path by a tooltip.

I imagine the shortest path would be to extend the docs system to work at the project-level with those tooltips (be able to effectively extend the built in documentation with project-level docs, including config property paths), but I don't actually know. Perhaps this is already possible in an opaque/undocumented way?

I'd be very interested in this.

**Reply by Calinou** [reply]
Descriptions are sourced from https://github.com/godotengine/godot/blob/master/doc/classes/EditorSettings.xml, which is statically compiled in the engine binary.

GDScript code can have its own documentation that is generated in real-time, but this isn't available in non-GDScript languages yet.

I suppose the easy way to go around this would be to add a `description` field as suggested, but it may be nontrivial to handle this from the editor side (handling descriptions from a second data source in a single class display).

PS: The above also applies to ProjectSettings.

**Reply by asciifaceman** [reply]
Would treating it like an override make sense? If it is defined in project docs, override the first data source.

**Reply by AnidemDex** [reply]
Maybe it can be used in object properties as well? Sometimes I want to add a description for custom properties added through `_get_property_list`

**Comment by betalars** [comment]
I would love to see this for the sake of improving project settings as well.

**Reply by neth392** [reply]
This is something that would greatly benefit addon/plugin creators, and the end users of them so that they don't need to refer to the project's github for every setting

**Comment by CycloneRing** [comment]
Please implement this for GDExtension as well.

**Comment by sylbeth** [comment]
Is this still not a thing? Why does the editor have so many things unavailable to users?

**Reply by Calinou** [reply]
This isn't implemented yet.

> Why does the editor have so many things unavailable to users?

As per the [best practices for engine contributors](https://docs.godotengine.org/en/stable/contributing/development/best_practices_for_engine_contributors.html), we try to avoid speculation when designing APIs. This leads to some things being hardcoded or not exposed to the scripting API at first, unless there's a proven need for it.

**Reply by sylbeth** [reply]
What are the hurdles for this to be implemented? I have only dabbled with the GDExtension icons, but I could get around to try and PR.

I understand the need not to create useless solutions to things no one uses, but I'd still say documentation is not an option but a must. Are there no guidances on the matter? It feels the same with access to editor icons with GDExtension nodes, but that may just be my naïvety talking as I am no maintainer of a project as big as this.

**Reply by Calinou** [reply]
I don't know what the API for exposing documentation for custom project/editor settings would look like, but if you're interested in implementing this, I suggest joining the [Godot Contributors Chat](https://chat.godotengine.org/) and asking in the `#editor` channel.

There's definitely interest in this feature, but we should make sure it's usable from all languages (which means it can't rely on GDScript/C#/GDExtension-specific mechanisms).

**Reply by sylbeth** [reply]
Alright, thank you. I will take a deep look at it when I have the time, since now I don't have it. So if anyone else wants to implement it and has the time, please do go ahead.

**Comment by PiCode9560** [comment]
For a temporary fix, I made a script that you can add to your plugin that allow you to do this. https://github.com/PiCode9560/Godot-Editor-Settings-Description
